ledger/participant-integration-api/src/main/resources/db/migration/h2database-appendonly/V1__Append_only_schema.sql

```diff
@@ -121,9 +121,18 @@ CREATE TABLE participant_command_completions (
     application_id VARCHAR NOT NULL,
     submitters ARRAY NOT NULL,
     command_id VARCHAR NOT NULL,
+    -- The transaction ID is `NULL` for rejected transactions.
     transaction_id VARCHAR,
-    status_code INTEGER,
-    status_message VARCHAR
+    -- The three columns below are `NULL` if the completion is for an accepted transaction.
+    -- The `rejection_status` contains a Protocol-Buffers-serialized message of type
+    -- `google.rpc.Status`, containing the code, message, and further details (decided by the ledger
```

Comment:
Instead of storing the full serialized `google.rpc.Status` message, another option worth considering is storing the contents of `Status.details` only in a serialized form (i.e., in addition to the existing `code` and `message` fields). This may be a good compromise between storing the full serialized status message vs. completely decoupling from it, as we don't have a type for the contents of the `details` field anyway.

Reply:
I like this suggestion.
Especially so that it would still make use of the old fields, so no worries with migration, maintaining dead compatibility code etc.
Also this covers @nicu-da 's concern with the dedicated, in-house protobuf schema
Also it would allow more easily to pass additional detail content for accpeted completion as well (not sure that is needed at all)

Reply:
That is a great suggestion. It will also simplify ad-hoc debugging in the IndexDB, as status_code and status_message are still accessible to queries.

Reply:
The price is probably a bit more logic but I'd also lean towards "blob == only things that _need_ to be opaque", if possible.

Reply:
I think this is reasonable. However, we'd need to create a new protobuf for the details, or a custom serialization mechanism, as it's a `repeated Any`, which cannot, AFAIK, be easily serialized and deserialized.

Another way of dealing with this might be to use `BYTEA ARRAY`, or just `ARRAY` in H2.

What do you think?

Reply:
I would advise to use a well-defined proto message with buf checks then.

Oracle does not seem to support `BYTEA ARRAY`: https://stackoverflow.com/questions/16885526/return-list-type-array-of-blobs-in-oracle-11g

Reply:
> Another way of dealing with this might be to use BYTEA ARRAY, or just ARRAY in H2.

From the modelling perspective, in this case an array looks more convenient to me and it doesn't really make the details less opaque. I am not sure that there aren't DB-related impacts though. @digital-asset/kv-participant?

Reply:
The downside of using a protobuf message is that it can't be easily embedded into a `Status` without deserialization and reserialization, so we're still paying that cost.

That said, I think it's still worth doing to see how it goes.

Reply:
Made a PR on top of this one to get a separate review. Please take a look at https://github.com/digital-asset/daml/pull/10613 and let me know if you approve.

Reply:
> The downside of using a protobuf message is that it can't be easily embedded into a Status without deserialization and reserialization, so we're still paying that cost.

I am sceptical about this aspect:
* To my best knowledge this is tiny data, so additional serialization effort will be also tiny
* Since it is null for all successful completion, we would optimize only the rejection case, which is also (hopefully) tiny fraction compared to successful completions

Reply:
OK, merged these changes into this PR branch. Let me know if you have any further concerns.